### DISCUSSION TITLE
Light weight integration tests

### DISCUSSION THREAD

**Discussion by firelizzard18** [golang/vscode-go#1789] [Discussions]
Would you (the Go team) be interested in (a PR) implementing integration tests that can be run directly, without spawning an instance of VSCode? The [GitLab extension](https://gitlab.com/gitlab-org/gitlab-vscode-extension) does this by creating a [partial mock of the `vscode` package](https://gitlab.com/gitlab-org/gitlab-vscode-extension/-/blob/main/src/__mocks__/vscode.js).

[the remark this answers is not in the document]
**Reply by hyangah** [reply]
What do you think about `typemoq`? I saw python extension and other extensions using it 
For example: https://github.com/microsoft/vscode-python/search?q=typemoq

[the remark this answers is not in the document]
**Reply by firelizzard18** [reply]
My goal is to allow (some) tests to be run without launching VSCode. Since `vscode` is normally provided by VSCode/the extension host, to allow tests to be run independently, we have to intercept `require('vscode')`. The only test framework I know of that supports intercepting `require` is Jest. I skimmed `typemoq`'s readme and I didn't see anything about `require` or `import`. I don't have any objections to `sinon` as a mocking framework (nor any objections to `typemoq`).

I played with an number of existing solutions for intercepting `require`. I think the simplest solution is to do what `mock-require` does: replace `Module._load` (from [this WIP CL](https://go-review.googlesource.com/c/vscode-go/+/351109)):

```typescript
import Module = require('module');

const { _load } = Module;

Module._load = function (request: string, parent?: Module, ...rest: any[]) {
	switch (request) {
		case 'vscode':
			return require('./mocks/MockVSCode');
		default:
			return _load.call(this, request, parent, ...rest);
	}
};

require('./integration/index').run();
```

This intercepts `require('vscode')` and replaces it with whatever code we want. The implementation is straight forward and easy to extend (we might want to intercept `vscode-languageclient`).

**However, there is an issue with respect to minimizing the mock `vscode`**: The import graph of/between files in `src` is quite interdependent/tangled, and as a result, `goMain` is almost always imported, no matter what you start with. If we want to go down this path, I recommend simplifying the intra-`src` import graph, so we don't need to create mocks for features that aren't tested.

For example, if I want to test `GoTestResolver`, I hit the following import chain
- `goTest/resolve`
- `goStatus`
- `goEnvironmentStatus`
- `goEnv`
- `util`
- `goLanguageServer`
- `goCodeAction`
- `goImport`
- `goInstallTools`
- `goMain`

`goMain` effectively loads the entire extension, which means *everything* the extension uses anywhere must be mocked. Additionally, loading `goStatus` requires mocking `OutputChannel` and `StatusBarItem`, and loading `goLanguageServer` requires intercepting `vscode-languageclient`. Alternatively, we could add a mechanism to intercept arbitrary modules (or just use `mock-require` or `proxyquire`) and use that to mock `require('./goStatus')`.
